pkg/reconciler/route/traffic/rollout.go

```diff
@@ -70,3 +72,94 @@ type RevisionRollout struct {
 	// target percentage.
 	Percent int `json:"percent"`
 }
+
+// Roll processes previous Rollout object and compares to the current
+// rollout state. If there is different, Merge will start or stop the rollout
+// and update `RevisionRollout` objects accoringly.
+// Roll returns true if any changes have been made.
+func (cur *Rollout) Roll(prev *Rollout) bool {
+	if prev == nil {
+		return false
+	}
+
+	// The algorithm below is simplest, but probably not the most performant.
+	// Optimize in the later passes.
+	// Map by tag.
+	ccfg, pcfg := map[string][]*ConfigurationRollout{}, map[string][]*ConfigurationRollout{}
+	for i := range cur.Configurations {
+		cfg := &cur.Configurations[i]
+		ccfg[cfg.Tag] = append(ccfg[cfg.Tag], cfg)
+	}
+	for i := range prev.Configurations {
+		cfg := &prev.Configurations[i]
+		pcfg[cfg.Tag] = append(pcfg[cfg.Tag], cfg)
+	}
+	ret := false
+	for t, cfgs := range ccfg {
+		pcfgs, ok := pcfg[t]
+		// A new tag, rollout to 100% so ignore.
+		if !ok {
+			continue
+		}
+		for i, j := 0, 0; i < len(cfgs) && j < len(pcfgs); {
+			switch {
+			case cfgs[i].ConfigurationName == pcfgs[j].ConfigurationName:
+				// Config might have 0 traffic assigned, if it is a tagged route.
+				if cfgs[i].Percent != 0 {
+					ret = ret || rollConfig(cfgs[i], pcfgs[j])
+				}
+				i++
+				j++
+			case cfgs[i].ConfigurationName < pcfgs[j].ConfigurationName:
+				// A new config, has been added. No action for rollout though.
+				i++
+			default: // cur > prev.
+				// A config has been removed. Again, no action for rollout, since
+				// this will no longer be rolling out even it were.
+				j++
+			}
+		}
+		// Non interlaping configs don't matter since we either roll 100% or remove them.
+	}
+	return ret
+}
+
+func rollConfig(cur, prev *ConfigurationRollout) bool {
+	pc := len(prev.Revisions)
+	// curr will always have just 1 element – the current desired revision.
```

Comment:
perhaps "goal" would be a better name than "current"?

Reply:
Sure. DOne.